contract_comm/types.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2021 The Celo Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package contract_comm
+
+import (
+	"math/big"
+	"reflect"
+
+	"github.com/celo-org/celo-blockchain/accounts/abi"
+	"github.com/celo-org/celo-blockchain/common"
+	ccerrors "github.com/celo-org/celo-blockchain/contract_comm/errors"
+	"github.com/celo-org/celo-blockchain/core/types"
+	"github.com/celo-org/celo-blockchain/core/vm"
+	"github.com/celo-org/celo-blockchain/log"
+)
+
+// SystemContractCaller runs core contract calls. The implementation provides the state on which the calls are ran
+type SystemContractCaller interface {
+	AddressCaller
+	RegistryCaller
+	GetRegisteredAddress(registryId common.Hash) (*common.Address, error)
+}
+
+// AddressCaller makes a call with a specified address
+type AddressCaller interface {
+	MakeStaticCallWithAddress(contractAddress common.Address, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64) (uint64, error)
+	MakeCallWithAddress(contractAddress common.Address, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int) (uint64, error)
+}
+
+// RegistryCaller looks up the contract address from the registry ID and then makes an address cal
+type RegistryCaller interface {
+	MakeStaticCall(registryId common.Hash, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64) (uint64, error)
+	MakeCall(registryId common.Hash, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int) (uint64, error)
+}
+
+// Creates a new EVM on demand.
+type evmBuilder interface {
+	// createEVM creates an EVM. May fail due to `chain.State()`
+	createEVM() (*vm.EVM, error)
+}
+
+// Private struct that implements SystemContractCaller.
+type contractCommunicator struct {
```

Comment:
I think we should keep the struct definition with the implementation, and would recommend we move this to `caller.go`